chain/client/src/stateless_validation/endorsement_tracker.rs

```diff
@@ -0,0 +1,141 @@
+    /// We first verify the chunk endorsement and then store it in a cache.
+    /// We would later include the endorsements in the block production.
+    fn process_chunk_endorsement(
+        &mut self,
+        chunk_header: ShardChunkHeader,
+        endorsement: ChunkEndorsement,
+    ) -> Result<(), Error> {
+        let chunk_hash = endorsement.chunk_hash();
+        let account_id = &endorsement.account_id;
+
+        // If we have already processed this chunk endorsement, return early.
+        if self
+            .chunk_endorsements
+            .get(chunk_hash)
+            .is_some_and(|existing_endorsements| existing_endorsements.get(account_id).is_some())
+        {
+            tracing::debug!(target: "stateless_validation", ?endorsement, "Already received chunk endorsement.");
+            return Ok(());
+        }
+
+        if !self.epoch_manager.verify_chunk_endorsement(&chunk_header, &endorsement)? {
+            tracing::error!(target: "stateless_validation", ?endorsement, "Invalid chunk endorsement.");
+            return Err(Error::InvalidChunkEndorsement);
+        }
+
+        // If we are the current block producer, we store the chunk endorsement for each chunk which
+        // would later be used during block production to check whether to include the chunk or not.
+        // TODO(stateless_validation): It's possible for a malicious validator to send endorsements
+        // for 100 unique chunks thus pushing out current valid endorsements from our cache.
+        // Maybe add check to ensure we don't accept endorsements from chunks already included in some block?
+        // Maybe add check to ensure we don't accept endorsements from chunks that have too old height_created?
+        tracing::debug!(target: "stateless_validation", ?endorsement, "Received and saved chunk endorsement.");
+        self.chunk_endorsements.get_or_insert(chunk_hash.clone(), || HashMap::new());
+        let chunk_endorsements = self.chunk_endorsements.get_mut(chunk_hash).unwrap();
+        chunk_endorsements.insert(account_id.clone(), endorsement);
+
+        Ok(())
+    }
+
+    /// Called by block producer.
+    /// Returns Some(signatures) if node has enough signed stake for the chunk represented by chunk_header.
+    /// Signatures have the same order as ordered_chunk_validators, thus ready to be included in a block as is.
+    /// Returns None if chunk doesn't have enough stake.
+    /// For older protocol version, we return an empty array of chunk endorsements.
+    pub fn get_chunk_endorsement_signatures(
+        &self,
+        chunk_header: &ShardChunkHeader,
+    ) -> Result<Option<ChunkEndorsementSignatures>, Error> {
+        let epoch_id =
+            self.epoch_manager.get_epoch_id_from_prev_block(chunk_header.prev_block_hash())?;
+        let protocol_version = self.epoch_manager.get_epoch_protocol_version(&epoch_id)?;
+        if !checked_feature!("stable", ChunkValidation, protocol_version) {
```

Comment:
We can also consider renaming ChunkValidation to StatelessValidation.

Reply:
StatelessValidationStage0 maybe? StatelessValidation is a very big name :) Also I still feel like this feature flag is currently only used for stateless chunk validation, not anything else we're planning to launch with stateless validation like validator role change, so maybe StatelessChunkValidation?

Reply:
StatelessValidation is a big name because we call the whole project that, not of itself ;) For the sake of naming things in code I wouldn't consider role chages etc. as stateless validation. Anyway ChunkValidation is confusing because we most certainly do already validate chunks today so any change would be good.

Reply:
Hmm, agree with "StatelessValidation" but not sure if we have consensus here? I'm leaving `ChunkValidation` as is for this PR and we can get back.